File: pointer/#125_valid_palindrome/isPalindrome.go
package main

import (
	"fmt"
	"regexp"
	"strings"
)

func main() {
	// isPalindrome("race a car")
	isPalindrome2("race a car")                     // false
	isPalindrome2("A man, a plan, a canal: Panama") // true
	isPalindrome2("0P")                             // false
}

func isPalindrome(s string) bool {
	regexp, _ := regexp.Compile(`[^a-zA-Z0-9]`)

	str := regexp.ReplaceAllString(s, "")
	str = strings.ToLower(str)
	for l, r := 0, len(str)-1; l < r; l, r = l+1, r-1 {
		if str[l] != str[r] {
			fmt.Println(false)
			return false
		}
	}

	fmt.Println(true)
	return true
}

func isPalindrome2(s string) bool {
	isAlphanum := func(b byte) bool {
		return (b >= 48 && b <= 57) || (b >= 65 && b <= 90) || (b >= 97 && b <= 122)
		//			  ^			^			  ^			^			  ^			 ^
		//ASCII Table '0'       '9'			 'A'		'Z'			 'a'		'z'
	}

	for l, r := 0, len(s)-1; l < r; l, r = l+1, r-1 {
		for l < r && !isAlphanum(s[l]) {
			l++
		}

		for l < r && !isAlphanum(s[r]) {
			r--
		}

		sl, sr := s[l], s[r]
		if sl >= 65 && sr >= 65 { // check if sl, sr is alpha char
			if sl > sr { // change char to lowercase if sl equal sr (EX: 'a' <=> 'A'), if doesn't mattar not equal, it will check at line57.
				sr += 32
			} else if sl < sr {
				sl += 32
			}
		}

		if sl != sr {
			fmt.Println(false)
			return false
		}
	}

	fmt.Println(true)
	return true
}

File: pointer/#125_valid_palindrome/isPalindrome.js
/**
 * A phrase is a palindrome if, after converting all uppercase letters into lowercase letters and removing all non-alphanumeric characters
 * @param {string} s
 * @return {boolean}
 */
var isPalindrome = function (s) {
  // 將a-z, A-Z, 0-9以外的字元移除
  let str = s.replaceAll(/[^a-zA-Z0-9]/g, "").toLowerCase();

  let l = 0,
    r = str.length - 1;
  while (l < r) {
    if (str[l] !== str[r]) {
      console.log(false);
      return false;
    }
    l++;
    r--;
  }

  console.log(true);
  return true;
};

isPalindrome("A man, a plan, a canal: Panama"); // true
isPalindrome("race a car"); // false
isPalindrome(" "); // true
isPalindrome("0P"); // false

// https://leetcode.com/problems/valid-palindrome/solutions/3133780/3-simple-javascript-solutions/
var isPalindrome2 = function (s) {
  const reg = new RegExp(/[^a-zA-Z0-9]/);

  let l = 0,
    r = s.length - 1;
  while (l < r) {
    while (l < r && reg.test(s[l])) l++;
    while (l < r && reg.test(s[r])) r--;
    if (s[l].toLowerCase() !== s[r].toLowerCase()) {
      return false;
    }
    l++;
    r--;
  }

  return true;
};
